src/sdi-notify.c

```diff
@@ -0,0 +1,278 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <gio/gdesktopappinfo.h>
+#include <glib/gi18n.h>
+
+#include "sdi-helpers.h"
+#include "sdi-notify.h"
+
+enum { PROP_APPLICATION = 1, PROP_LAST };
+
+struct _SdiNotify {
+  GObject parent_instance;
+
+  GApplication *application;
+};
+
+G_DEFINE_TYPE(SdiNotify, sdi_notify, G_TYPE_OBJECT)
+
+static GTimeSpan get_remaining_time(SnapdSnap *snap) {
+  g_autoptr(GDateTime) proceed_time = snapd_snap_get_proceed_time(snap);
+  g_autoptr(GDateTime) now = g_date_time_new_now_local();
+  return g_date_time_difference(proceed_time, now);
+}
+
+static void show_pending_update_notification(SdiNotify *self,
+                                             const gchar *title,
+                                             const gchar *body, GIcon *icon,
+                                             GSList *snaps) {
+
+  g_autoptr(GNotification) notification = g_notification_new(title);
+  g_notification_set_body(notification, body);
+  if (icon != NULL) {
+    g_notification_set_icon(notification, g_object_ref(icon));
+  }
+  g_notification_set_default_action_and_target(
+      notification, "app.close-notification", "s", "pending-update");
+  g_notification_add_button_with_target(
+      notification, "Close", "app.close-notification", "s", "pending-update");
+  g_autoptr(GVariantBuilder) builder =
+      g_variant_builder_new(G_VARIANT_TYPE("as"));
+  for (; snaps != NULL; snaps = snaps->next) {
+    SnapdSnap *snap = (SnapdSnap *)snaps->data;
+    g_variant_builder_add(builder, "s", snapd_snap_get_name(snap));
+  }
+  GVariant *values = g_variant_builder_end(builder);
+  g_notification_add_button_with_target_value(notification, "Ignore",
+                                              "app.ignore-updates", values);
+  g_application_send_notification(self->application, "pending-update",
+                                  notification);
+}
+
+static void show_simple_notification(SdiNotify *self, const gchar *title,
+                                     const gchar *body, GIcon *icon,
+                                     const gchar *ID) {
+  g_autoptr(GNotification) notification = g_notification_new(title);
+  g_notification_set_body(notification, body);
+  if (icon != NULL) {
+    g_notification_set_icon(notification, g_object_ref(icon));
+  }
+  g_application_send_notification(self->application, ID, notification);
+}
+
+void sdi_notify_pending_refresh_one(SdiNotify *self, SnapdSnap *snap) {
+  g_return_if_fail(SDI_IS_NOTIFY(self));
+  g_return_if_fail(snap != NULL);
+
+  GIcon *icon = NULL;
+  g_autofree gchar *body = NULL;
```

Comment:
Nit: these should be just above where they are set.

Reply:
Done.